tests/test_actions.py

```diff
@@ -479,48 +522,137 @@ def test_restore_action(self):
+        tls_provider.all_requests = [tls_cert_relation]
+        actions.endpoint_from_name.return_value = tls_provider
+        actions.create_client_certificate.return_value = (client_cert,
+                                                          client_key)
 
         # Generate client certificate for host missing in backup
         # builtin 'open()' function will raise FileNotFoundError, which acts
         # as if the file was not found in the backup
         with patch('builtins.open', new_callable=mock_open, read_data='data') \
                 as mock_file:
-            mock_file.side_effect = FileNotFoundError()
+            mock_file.side_effect = FileNotFoundError
             with self.func_call_arguments(name='backup.tar.gz'):
                 self.call_action()
 
-                actions._verify_backup.assert_called()
-                actions._replace_pki.assert_called()
-                actions._update_leadership_data.assert_called()
-                tls_provider.set_ca.assert_called()
-                tls_provider.set_client_cert.assert_called()
-                actions.create_client_certificate.assert_called()
-                tls_cert_relation.set_cert.assert_called()
+        self.assert_common_restore_actions(tls_provider)
+        actions.create_client_certificate.assert_called_once_with(
+            tls_cert_relation.common_name)
+        tls_cert_relation.set_cert.assert_called_once_with(client_cert,
+                                                           client_key)
+
+    def test_restore_action_server_missing(self):
+        """Test 'restore' action when new server cert needs to be generated.
+
+        This scenario happens when server certificate for currently connected
+        unit is not found in the backup bundle.
+        """
+        mock_internal = {
+            actions: ['_verify_backup',
+                      '_replace_pki',
+                      '_update_leadership_data',
+                      'create_client_certificate',
+                      'create_server_certificate',
+                      ]
+        }
+        self.patch_all(mock_internal)
+
+        server_cert = MagicMock()
+        server_key = MagicMock()
+        tls_provider = MagicMock()
+        tls_cert_relation = tls_certificate_relation('tls_server', 'server')
+        tls_provider.all_requests = [tls_cert_relation]
+        actions.endpoint_from_name.return_value = tls_provider
+        actions.create_server_certificate.return_value = (server_cert,
+                                                          server_key)
 
         # Generate server certificate for host missing in backup
-        tls_server_relation = tls_certificate_relation('tls_server',
-                                                       'server')
-        tls_provider.all_requests = [tls_server_relation]
+        # builtin 'open()' function will raise FileNotFoundError, which acts
```

Comment:
When I tried to run:
```python
    with self.func_call_arguments(name='not-valid-backup.tar.gz'):
        self.call_action()
        self.assert_function_fail_msg("File not found")
```
I fount out that `function_fail` was called with arguments:
```python
[call('Action restore failed: Unrecognized certificate request type "application".')]
``` 
Is this ok?

Reply:
Hmm..when I change filename and add the `assert_function_fail_msg("FIle not found")` I get
```
E           AssertionError: expected call not found.
E           Expected: mock('Action restore failed: File not found')
E           Actual: not called.
```
Which is expected, action should not fail in this UT. Changing name of the "backup file" in `func_call_arguments` does not really matter, to simulate missing backup file, you'd need to set `os.path.isfile` return value to `False`, but there's already test for that (`test_missing_backup_file`)

Reply:
> Hmm..when I change filename and add the `assert_function_fail_msg("FIle not found")` I get
> 
> ```
> E           AssertionError: expected call not found.
> E           Expected: mock('Action restore failed: File not found')
> E           Actual: not called.
> ```

Same here, however I printed all calls and there was only `[call('Action restore failed: Unrecognized certificate request type "application".')]`, which seems little bit odd, because I expected something like `file not found`.

> Which is expected, action should not fail in this UT. Changing name of the "backup file" in `func_call_arguments` does not really matter, to simulate missing backup file, you'd need to set `os.path.isfile` return value to `False`, but there's already test for that (`test_missing_backup_file`)

Oh well, I get it. Thanks for the explanation.

Reply:
Ok, good, that's expected then.